proposals/0008-uninitialized-resources.md

```diff
@@ -0,0 +1,116 @@
+code generation), and changes the rules around resource validation.
+
+## Motivation
+
+The current implementation has two key problems:
+
+* Quality of diagnostic locations degrades post-optimization
+* Seemingly innocuous code changes make code invalid
+
+As an example take the following invalid code:
+
+```c++
+RWBuffer<int> In;
+RWBuffer<int> Out[2];
+
+void fn(bool Cond) {
+  RWBuffer<int> O;
+  if (Cond)
+    O = Out[1];
+  O[0] = In[0]; // error: local resource not guaranteed to map to unique global resource.
+}
+
+[numthreads(1,1,1)]
+void main() {
+  fn(false);
+}
+```
+(see: https://godbolt.org/z/KG5v98s8b)
+
+Note, the comment indicating where the diagnostic is emitted. Because the
+analysis is performed after optimization, it is only able to materialize a
+location for the use of the unresolved value. In this trivial case it is easy
+enough to see where the error lies, but in a more complicated example it might
+not be so apparent.
+
+Further, this code compiles successfully if the boolean value passed to `fn` is
+`true` instead of `false`. This demonstrates the ability for seemingly unrelated
+code changes to impact code correctness.
+
+## Proposed solution
+
+Building off clang's existing support for uninitialized variable analysis, we
+can move this analysis into the Clang control flow graph (CFG) instead of post
+optimization, which enables more robust diagnostic generation.
+
+Take the following alternate diagnostics:
+
+```c++
+RWBuffer<int> In;
+RWBuffer<int> Out[2];
+
+void fn(bool Cond) {
+  RWBuffer<int> O; // note: variable 'O' is declared here
+  if (Cond) // warning: variable 'O' is used uninitialized whenever 'if' condition is false}}
+    O = Out[1];
+  O[0] = In[0]; // note: uninitialized use occurs here
+}
+
+[numthreads(1,1,1)]
+void main() {
+  fn(false);
+}
+```
+
+These diagnostics already exist and are emitted by Clang for uninitialized
+variable usage, and provide more actionable and understandable feedback to the
+user.
+
+In addition to generating these diagnostics using Clang's CFG, we should make
+uninitialized use, or potential uninitialized use, of resource variables an
+error during compilation. Making this always an error removes the ability for
+seemingly unrelated changes to cascade resulting in unexpected compiler errors.
+
+A potential starting point for an implementation of this change is posted on
+LLVM's [Phabricator](https://reviews.llvm.org/D130055).
+
+A complete implementation requires an IPC analysis to ensure that any code paths
+that initialize a local resource based on a parameter can determine the value of
+the parameter at compile time.
+
+For example:
+
+```c++
+RWBuffer<int> In;
+RWBuffer<int> Out[2];
+void fn(bool Cond) {
+  RWBuffer<int> O = Out[0]; // warning: local resource not guaranteed to resolve to a unique global resource.
+  if (Cond) // note: if statement here is not compile-time resolved
+    O = Out[1];
+  O[0] = In[0]; // note: unresolved use occurs here

```

Comment:
Technically, since O is always assigned to a single resource range declaration, this is (can be made) legal.  But there is a problem if `Cond` is non-uniform, since NonUniformResourceIndex would need to be used at the indexing (create handle) site, and that indexing would be created by the compiler when replacing the conditional assignment of a resource with a conditional assignment of an index.

If instead of `Out[2]` you had separate `Out1` and `Out2` variables, then used those instead of `Out[0]` and `Out[1]`, then it would require `Cond` to be compile-time resolved.